### PR TITLE
feat(@aws-amplify/pushnotification) provide config option for not automatically requesting iOS permissions per #1781

### PR DESCRIPTION
**Pull request by lafiosca**
*Issue #, if available:* #1781

*Description of changes:* Per discussion in the issue thread, this PR provides a new config option `requestIOSPermissions` for the `PushNotification` config. This is a boolean flag which specifies whether or not to automatically request push notifications permission in iOS when calling `PushNotification.configure` for the first time. The option defaults to `true` if not provided, to maintain backward compatibility. The PR also includes a new method `PushNotification.requestIOSPermissions` for explicitly requesting the permissions at the desired moment in the app flow (e.g., after a user has logged in). The PR also includes some minor semicolon-related style fixes to satisfy lint.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by kbruck-rsc** [comment]
My company needs this as well. Can we get this merged in ASAP???

**Comment by silviopaganini** [comment]
Need this too! any updates?

**Comment by joebernard** [comment]
@powerful23 Can we please get some feedback on this PR? Not being able to control when the user is prompted for notifications permission is a major blocker for us. Can you provide some insight here?

File: packages/pushnotification/src/PushNotification.ts [hunk 1]
```diff
@@ -61,14 +61,20 @@ export default class PushNotification {
...
+            conf,
+        );
 
         if (Platform.OS === 'android' && !this._androidInitialized){
             this.initializeAndroid();
             this._androidInitialized = true;
         } 
         else if (Platform.OS === 'ios' && !this._iosInitialized) {
-            this.initializeIOS();
+            this.initializeIOS(this._config.requestIOSPermissions);
```
Comment (powerful23) [inline packages/pushnotification/src/PushNotification.ts]: We don't really need to pass the configuration down into a private function.
```js
            this.initializeIOS();
```

File: packages/pushnotification/src/PushNotification.ts [hunk 2]
```diff
@@ -111,12 +117,18 @@ export default class PushNotification {
         RNPushNotification.initialize();
     }
 
-    initializeIOS() {
+    requestIOSPermissions() {
```
Comment (powerful23) [inline packages/pushnotification/src/PushNotification.ts]: Maybe it's better to give the control of specify the parameters of this function like
```js
    requestIOSPermissions(options={ alert:true, badge: true, sound: true }) {
          PushNotificationIOS.requestPermissions(options);
    }
```

File: packages/pushnotification/src/PushNotification.ts [hunk 3]
```diff
@@ -111,12 +117,18 @@ export default class PushNotification {
...
         PushNotificationIOS.requestPermissions({
             alert: true,
             badge: true,
             sound: true
         });
+    }
+
+    initializeIOS(requestPermissions: boolean = true) {
+        if (requestPermissions) {
```
Comment (powerful23) [inline packages/pushnotification/src/PushNotification.ts]:
```js
if (this._config.requestIOSPermissions) this.requestIOSPermissions();
```

File: packages/pushnotification/src/PushNotification.ts (hunk 1 above)
Comment (lafiosca) [reply packages/pushnotification/src/PushNotification.ts]: Updated

File: packages/pushnotification/src/PushNotification.ts (hunk 2 above)
Comment (lafiosca) [reply packages/pushnotification/src/PushNotification.ts]: Updated

File: packages/pushnotification/src/PushNotification.ts (hunk 3 above)
Comment (lafiosca) [reply packages/pushnotification/src/PushNotification.ts]: Updated

**Comment by lafiosca** [comment]
@powerful23 I've pushed the requested changes to my branch. I will look into submitting a PR for the docs. I guess I should just add a little aside to the PushNotifications "Configure your App" section?

**Comment by lafiosca** [comment]
@powerful23 Documentation PR added here: https://github.com/aws-amplify/docs/pull/643

**Comment by joebernard** [comment]
@powerful23 Are there any further changes needed to get this merged?

**Comment by powerful23** [comment]
@lafiosca Great, thanks for your contribution!

**Comment by joebernard** [comment]
`requestIOSPermissions` seems to have stopped working in recent versions. I get the immediate prompt for notification permissions regardless of this setting.

```
"@aws-amplify/core": "^1.2.4",
"@aws-amplify/pushnotification": "^1.1.4",
"react-native": "0.61.3",
```

I'm using the built in deprecated RN push notifications (not the react-native-community version) because of the issues documented in #4089 .

**Pull request merged**